diff/diff.go

```diff
@@ -0,0 +1,293 @@
+package diff
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// Difference represents a "chunk" of difference between two
+// `Object`s.
+type Difference interface {
+	Summarise(out io.Writer)
+	AtPath() string
+}
+
+type Chunk struct {
+	Deleted []interface{}
+	Added   []interface{}
+	Path    string
+}
+
+func (c Chunk) AtPath() string {
+	return c.Path
+}
+
+// Some convenience funcs for when a single value has changed,
+// appeared or disappeared.
+
+func Changed(A, B interface{}, path string) Difference {
+	return Chunk{
+		Path:    path,
+		Deleted: []interface{}{A},
+		Added:   []interface{}{B},
+	}
+}
+
+func Added(B interface{}, path string) Difference {
+	return Chunk{
+		Path:  path,
+		Added: []interface{}{B},
+	}
+}
+
+func Removed(A interface{}, path string) Difference {
+	return Chunk{
+		Path:    path,
+		Deleted: []interface{}{A},
+	}
+}
+
+// a value has changed, but don't report the before or after
+type OpaqueChunk struct {
+	Path string
+}
+
+func (c OpaqueChunk) AtPath() string {
+	return c.Path
+}
+
+// Objects are considered unique by {Namespace, Kind, Name}
+type ObjectID struct {
```

Comment:
General question. What happens if the user alters the namespace or name in the repository (unlikely to happen in the cluster)?
Then all elements in the repo will be marked as repo only, and the same for the cluster, but cluster only.
Probably fine.